### PR TITLE
[mlir][gpu] Add optional attributes of kernelModule and kernelFunc for outlining kernels.

### PR DESCRIPTION
**Pull request by wangzpgi**
Adding optional attributes so we can specify the kernel function names and the kernel module names generated.

### CODE REVIEW

**Comment by joker-eph** [comment]
Thanks for the patch, can you please add tests?

**Comment by clementval** [comment]
> Thanks for the patch, can you please add tests?

Thabks for the patch Zhen. As Mehdi mentioned, please add tests that cover all the use cases you expect.

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 1]
```diff
@@ -395,16 +403,35 @@ class GpuKernelOutliningPass
...
     // prevent manual building of Ops with symbols in code using SymbolTables
     // and then this needs to use the OpBuilder.
     auto *context = getOperation().getContext();
     OpBuilder builder(context);
-    auto kernelModule = builder.create<gpu::GPUModuleOp>(kernelFunc.getLoc(),
-                                                         kernelFunc.getName());
+    std::string kernelModuleName;
+    if (gpuLaunchOp.hasKernelModuleName()) {
+      kernelModuleName =
+          gpuLaunchOp->getAttrOfType<mlir::SymbolRefAttr>("kernelModule")
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:418]: You should be able to use the generated getter for the attributes

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 2]
```diff
@@ -364,17 +364,25 @@ class GpuKernelOutliningPass
       Block::iterator insertPt(func->getNextNode());
       auto funcWalkResult = func.walk([&](gpu::LaunchOp op) {
         SetVector<Value> operands;
-        std::string kernelFnName =
-            Twine(op->getParentOfType<SymbolOpInterface>().getName(), "_kernel")
-                .str();
+        std::string kernelFnName;
+        if (op.hasKernelFuncName()) {
+          kernelFnName = op->getAttrOfType<mlir::SymbolRefAttr>("kernelFunc")
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:369]: Why not using the generated getter?

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 3]
```diff
@@ -395,16 +403,35 @@ class GpuKernelOutliningPass
...
+    gpu::GPUModuleOp kernelModule;
+    // Check if the module already exists in the symbol table
+    if (auto existingModule =
+            parentSymbolTable.lookup<gpu::GPUModuleOp>(kernelModuleName)) {
+      kernelModule = existingModule;
+    } else {
+      // If not found, create a new GPU module
+      kernelModule = builder.create<gpu::GPUModuleOp>(kernelFunc.getLoc(),
+                                                      kernelModuleName);
+    }
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:434]:
```suggestion
    // Check if the module already exists in the symbol table
    gpu::GPUModuleOp kernelModule = parentSymbolTable.lookup<gpu::GPUModuleOp>(kernelModuleName);
    if (!kernelModule) {
      // If not found, create a new GPU module
      kernelModule = builder.create<gpu::GPUModuleOp>(kernelFunc.getLoc(),
                                                      kernelModuleName);
    }
```

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 4]
```diff
@@ -364,17 +364,25 @@ class GpuKernelOutliningPass
       Block::iterator insertPt(func->getNextNode());
       auto funcWalkResult = func.walk([&](gpu::LaunchOp op) {
         SetVector<Value> operands;
-        std::string kernelFnName =
-            Twine(op->getParentOfType<SymbolOpInterface>().getName(), "_kernel")
-                .str();
+        std::string kernelFnName;
+        if (op.hasKernelFuncName()) {
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:368]:
```suggestion
        if (op.getKernelFunc()) {
```

File: mlir/include/mlir/Dialect/GPU/IR/GPUOps.td [hunk 5]
```diff
@@ -1030,6 +1036,16 @@ def GPU_LaunchOp : GPU_Op<"launch", [
...
+
+    /// Checks if the kernel func name attribute is present.
+    bool hasKernelFuncName() {
+      return getKernelFunc().has_value();
+    }
+
+    /// Checks if the kernel module name attribute is present.
+    bool hasKernelModuleName() {
+      return getKernelModule().has_value();
+    }
```
Comment (clementval) [inline mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:1048]: You don't really need this. You can check optional like 

```
if (op.getKernelFunc()) {
  // attribute is present. 
}
```

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 6]
```diff
@@ -395,16 +401,34 @@ class GpuKernelOutliningPass
...
     // TODO: This code cannot use an OpBuilder because it must be inserted into
     // a SymbolTable by the caller. SymbolTable needs to be refactored to
     // prevent manual building of Ops with symbols in code using SymbolTables
     // and then this needs to use the OpBuilder.
     auto *context = getOperation().getContext();
     OpBuilder builder(context);
-    auto kernelModule = builder.create<gpu::GPUModuleOp>(kernelFunc.getLoc(),
-                                                         kernelFunc.getName());
+    std::string kernelModuleName;
+    gpu::GPUModuleOp kernelModule = nullptr;
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:414]:
```suggestion
    gpu::GPUModuleOp kernelModule;
```

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 7]
```diff
@@ -395,16 +401,34 @@ class GpuKernelOutliningPass
...
+    if (gpuLaunchOp.getKernelModule()) {
+      kernelModuleName =
+          gpuLaunchOp.getKernelModule()->getRootReference().str();
+      if (auto existingModule =
+              parentSymbolTable.lookup<gpu::GPUModuleOp>(kernelModuleName)) {
+        kernelModule = existingModule;
+      }
```
Comment (clementval) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:421]:
```suggestion
        kernelModule = parentSymbolTable.lookup<gpu::GPUModuleOp>(kernelModuleName);
```

File: mlir/test/Dialect/GPU/outlining.mlir [hunk 8]
```diff
@@ -508,3 +508,24 @@ func.func @launch_cluster() {
...
+  %bDimY = arith.constant 16 : index
+  %bDimZ = arith.constant 8 : index
+
+  gpu.launch blocks(%bx, %by, %bz) in (%grid_x = %gDimX, %grid_y = %gDimY, %grid_z = %gDimZ)
+             threads(%tx, %ty, %tz) in (%block_x = %bDimX, %block_y = %bDimY, %block_z = %bDimZ) {
+    "some_op"(%bx, %tx) : (index, index) -> ()
+    gpu.terminator
+  } {kernelModule = @test_module, kernelFunc = @test_kernel_func}
+  return
+}
```
Comment (clementval) [inline mlir/test/Dialect/GPU/outlining.mlir:531]: Can you add a test where the module is already present? Maybe also a test where only module is specified and one where only kernel name is specified to cover more use cases.

File: mlir/include/mlir/Dialect/GPU/IR/GPUOps.td [hunk 9]
```diff
@@ -837,6 +839,9 @@ def GPU_LaunchOp : GPU_Op<"launch", [
     -   a variadic number of Workgroup memory attributions.
     -   a variadic number of Private memory attributions.
 
+    The `kernelFunc` and `kernelModule` attributes are optional and specifies
+    the kernel name and a module in whichthe kernel should be outlined.
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:843]:
```suggestion
    the kernel name and a module in which the kernel should be outlined. 
```

**Review by clementval — APPROVED** [review]
LGTM! Thanks for addressing the comments. Maybe wait a bit to see if someone else has comment before merging.

**Review by razvanlupusoru — APPROVED** [review]
Thanks! Nice work!

**Review by Renaud-K — APPROVED** [review]

**Pull request merged**